arch/xtensa/src/common/xtensa_initialstate.c

```diff
@@ -100,6 +124,71 @@ void up_initial_state(struct tcb_s *tcb)
   xcp->regs[REG_A1] = (uint32_t)tcb->stack_base_ptr + /* Physical top of stack frame    */
                                 tcb->adj_stack_size;
 
+  /* Each task access the TLS variables using the THREADPTR register plus an
+   * offset to obtain the address of the variable.
+   *
+   * TLS layout at link-time, where 0xNNN is the offset that the linker
+   * calculates to a particular TLS variable.
+   *
+   * LOW ADDRESS
+   *         |---------------------------|   Linker Symbols
+   *         | Section                   |   --------------
+   *         | .flash.rodata             |
+   *      0x0|---------------------------| <- _rodata_reserved_start
+   *       ^ | Other Data                |
+   *       | |---------------------------| <- _thread_local_start
+   *       | | .tbss                     | ^
+   *       V |                           | |
+   *   0xNNN | int example;              | | tls_area_size
+   *         |                           | |
+   *         | .tdata                    | V
+   *         |---------------------------| <- _thread_local_end
+   *         | Other data                |
+   *         | ...                       |
+   *         |---------------------------|
+   * HIGH ADDRESS
+   *
+   * Calculate the THREADPTR register's initialization value based on the
+   * link-time offset and the TLS area allocated on the stack.
+   *
+   * HIGH ADDRESS
```

Comment:
it's a bit confusing to have to figures with the direction of address swapped...

Reply:
Yes, I agree... but it makes sense since the first figure is about flash and the second about the memory on the stack (and the stack grows towards lower memory addresses). I'll add a note about it to make it clearer.